### PR TITLE
Announce Accept Char id

### PR DESCRIPTION
**Pull request by sader1992**
* **Addressed Issue(s)**: https://github.com/rathena/rathena/issues/3901

* **Server Mode**: both

* **Description of Pull Request**: 
the original idea is to make a new GID like `GID_GROUP` and hook it to vector of players
but instead of making things complicated
this will allow `<char_id>` in the `announce` command as it already accept attached char on `bc_self`

* with this PR we can do this:
```cs
OnInit:
	bindatcmd("listenpvp",strnpcinfo(3)+"::Onlistenpvp",0,99);
end;

Onlistenpvp:
	.@ndx = inarray($pvplisteners,getcharid(0));
	if(.@ndx == -1){
		$pvplisteners[getarraysize($pvplisteners)] = getcharid(0);
		dispbottom "PVP Announcement : ON";
	}else{
		deletearray $pvplisteners[.@ndx],1;
		dispbottom "PVP Announcement : OFF";
	}
end;
OnSomethingHappen:
	for(.@i=0;.@i<getarraysize($pvplisteners);.@i++){
		announce "Something happen in PVP!",bc_self,0xFFF618,FW_NORMAL,12,NULL,NULL,$pvplisteners[.@i];
	}
end;
```

* **test script**:
```cs
//with both online char id 150000 , 150003
prontera,157,170,4	script	announce_to_id	444,{
	announce "this for the char id 150000",bc_self,0xFFF618,FW_NORMAL,12,NULL,NULL,150000;
	announce "this for the char id 150003",bc_self,0xFFF618,FW_NORMAL,12,NULL,NULL,150003;
}
```

### CODE REVIEW

**Comment by Litro** [comment]
how much loop with out sleep it will need freeloop ? 
```
OnSomethingHappen:
	for(.@i=0;.@i<getarraysize($pvplisteners);.@i++){
		announce "Something happen in PVP!",bc_self,0xFFF618,FW_NORMAL,12,NULL,NULL,$pvplisteners[.@i];
	}
end;
```
can this go like just send array name and server will handle the loop
```
OnSomethingHappen:
	announce "Something happen in PVP!",bc_self,0xFFF618,FW_NORMAL,12,NULL,NULL,$pvplisteners;
end;
```

**Comment by sader1992** [comment]
@Litro 
nope , the command accept int char id , it wont pass the reference of the array

for the free loop , i don't think there is a static number , in general as long as your array is above 50 use a free loop
something like this
```cs
OnSomethingHappen:
	.@size = getarraysize($pvplisteners);
	if(.@size > 50)
		freeloop(1);
	for(.@i=0;.@i<.@size;.@i++){
		announce "Something happen in PVP!",bc_self,0xFFF618,FW_NORMAL,12,NULL,NULL,$pvplisteners[.@i];
	}
	if(.@size > 50)
		freeloop(0);
end;
```

File: doc/script_commands.txt [hunk 1]
```diff
@@ -6856,6 +6857,9 @@ but it can be used instead in NPCs to "preview" an announce.
 	// This will be shown on everyones screen that is in sight of the NPC.
 	announce "This is my message just for you people here",bc_npc|bc_area;
 
+	// This will be a private message to the player with character id 150000
+	announce "This is my message just for char id 150000",bc_self,0xFFF618,FW_NORMAL,12,NULL,NULL,150000;
```
Comment (Atemo) [inline doc/script_commands.txt]: `NULL` hasn't been defined as constant (yet!)
```suggestion
	announce "This is my message just for char id 150000",bc_self,0xFFF618,FW_NORMAL,12,0,0,150000;
```

File: src/map/script.cpp [hunk 2]
```diff
@@ -11103,10 +11103,10 @@ BUILDIN_FUNC(announce)
 		}else{
 			struct map_session_data* sd;
 
-			if( script_rid2sd(sd) )
-				bl = &sd->bl;
-			else
+			if (!script_charid2sd(9, sd))
```
Comment (Lemongrass3110) [inline src/map/script.cpp]: Can you restore the original order here? Inverting the "if" adds a useless "not"-operation and actually all you do is exchange the called function from script_rid2sd to script_charid2sd.

**Pull request merged**